src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java

```diff
@@ -5337,10 +5338,15 @@ void writeRPCDateTime(String sName, GregorianCalendar cal, int subSecondNanos,
                 TDS.BASE_YEAR_1900);
 
         // Next, figure out the number of milliseconds since midnight of the current day.
-        int millisSinceMidnight = (subSecondNanos + Nanos.PER_MILLISECOND / 2) / Nanos.PER_MILLISECOND + // Millis into
-                                                                                                         // the current
-                                                                                                         // second
-                1000 * cal.get(Calendar.SECOND) + // Seconds into the current minute
+        int millisSinceMidnight;
+        // Millis into the current second
+        if (con.getServerMajorVersion() >= SQL_SERVER_VERSION_2016) {
+            millisSinceMidnight = getRoundedMilliseconds(subSecondNanos);
+        } else {
+            millisSinceMidnight = (subSecondNanos + Nanos.PER_MILLISECOND / 2) / Nanos.PER_MILLISECOND;
+        }
+
+        millisSinceMidnight = millisSinceMidnight + 1000 * cal.get(Calendar.SECOND) + // Seconds into the current minute
```

Comment:
it's nice to list out the calculations so it's easier to understand, but I'm wondering if we should just do all the calculations of the constant values 1st and just put this into the comments so we don't need to do this every time for performance reason?